File: frontend/src/api/minio.ts
import request from "@/utils/request"; // 根据实际项目路径调整
import type {
    UploadResponse,
    FileUrlResponse,
    FileListResponse,
} from "@/types/minio";

/**
 * 上传文件到 MinIO
 * @param formData 包含文件的 FormData
 * @param filename 可选，指定自定义文件名（允许覆盖）
 */
export function uploadFile(
    formData: FormData,
    filename?: string
): Promise<UploadResponse> {
    return request.post("/minio/upload", formData, {
        headers: { "Content-Type": "multipart/form-data" },
        params: filename ? { filename } : {},
    });
}

/**
 * 获取文件临时签名下载链接
 * @param key 文件存储 key
 * @param expires 有效期秒数，默认 3600
 */
export function getFileUrl(
    key: string,
    expires = 3600
): Promise<FileUrlResponse> {
    return request.get(`/minio/url/${encodeURIComponent(key)}`, {
        params: { expires },
    });
}

/**
 * 删除文件
 * @param key 文件存储 key
 */
export function deleteFile(key: string): Promise<void> {
    return request.delete(`/minio/delete/${encodeURIComponent(key)}`);
}

/**
 * 列出指定前缀下的文件
 * @param prefix 前缀路径（如 "approval"）
 */
export function listFiles(prefix: string): Promise<FileListResponse> {
    return request.get("/minio/list", {
        params: { prefix },
    });
}

/**
 * 批量获取文件 URL（并发）
 * @param keys 文件 key 数组
 * @param expires 签名有效期（秒）
 */
export async function batchGetFileUrls(
    keys: string[],
    expires = 3600
): Promise<Map<string, string>> {
    const map = new Map<string, string>();
    const promises = keys.map(async (key) => {
        try {
            const res = await getFileUrl(key, expires);
            map.set(key, res.url);
        } catch (err) {
            console.error(`获取 ${key} URL 失败`, err);
        }
    });
    await Promise.allSettled(promises);
    return map;
}

File: backend/src/controller/minio.rs
// handler/minio_handler.rs
use actix_multipart::Multipart;
use actix_web::{web, HttpResponse};
use futures::{TryStreamExt, StreamExt};
use bytes::BytesMut;
use serde::Deserialize;
use serde_json::json;
use std::collections::HashMap;
use actix_web::web::Query;
use crate::error::{AppError, AppResult};
use crate::service::ServiceContainer;
use crate::utils::response::success_response;

#[derive(Deserialize)]
pub struct UploadQuery {
    pub filename: Option<String>,
}

/// 上传文件
/// - 若提供 ?filename=xxx，则按 approval/xxx 存储，覆盖同名文件
/// - 否则使用 UUID 随机文件名
pub async fn upload_file(
    services: web::Data<ServiceContainer>,
    query: Query<UploadQuery>,
    mut payload: Multipart,
) -> AppResult<HttpResponse> {
    let mut field = payload.try_next().await
        .map_err(|e| AppError::InternalError(format!("读取multipart失败: {}", e)))?
        .ok_or_else(|| AppError::InvalidParameter("没有文件".into()))?;

    let content_type = field.content_disposition()
        .ok_or_else(|| AppError::InvalidParameter("缺少Content-Disposition".into()))?;
    let original_name = content_type.get_filename().unwrap_or("unknown").to_string();

    let mut bytes = BytesMut::new();
    while let Some(chunk) = field.next().await {
        let data = chunk.map_err(|e| AppError::InternalError(format!("读取文件块失败: {}", e)))?;
        bytes.extend_from_slice(&data);
    }

    let prefix = "approval"; // 所有文件统一放在此目录下

    let key = if let Some(custom_name) = &query.filename {
        // 使用自定义文件名（覆盖）
        let key = format!("{}/{}", prefix, custom_name);
        services.minio.upload_with_key(bytes.freeze(), &key).await
            .map_err(|e| AppError::InternalError(e.to_string()))?;
        key
    } else {
        // 回退到UUID随机文件名
        services.minio.upload(bytes.freeze(), &original_name, prefix).await
            .map_err(|e| AppError::InternalError(e.to_string()))?
    };

    Ok(success_response(json!({ "key": key })))
}

/// 获取文件下载URL
pub async fn get_file_url(
    services: web::Data<ServiceContainer>,
    path: web::Path<String>,
) -> AppResult<HttpResponse> {
    let key = path.into_inner();
    let url = services.minio.get_url(&key, 3600).await
        .map_err(|e| AppError::InternalError(e.to_string()))?;
    Ok(success_response(json!({ "url": url })))
}

/// 删除文件
pub async fn delete_file(
    services: web::Data<ServiceContainer>,
    path: web::Path<String>,
) -> AppResult<HttpResponse> {
    let key = path.into_inner();
    services.minio.delete(&key).await
        .map_err(|e| AppError::InternalError(e.to_string()))?;
    Ok(success_response(()))
}

/// 列出指定前缀的文件列表
pub async fn list_files(
    services: web::Data<ServiceContainer>,
    query: web::Query<HashMap<String, String>>,
) -> AppResult<HttpResponse> {
    let prefix = query.get("prefix").cloned().unwrap_or_else(|| "".to_string());
    let objects = services.minio.list_objects(&prefix).await
        .map_err(|e| AppError::InternalError(e.to_string()))?;

    let file_list: Vec<serde_json::Value> = objects.iter().map(|obj| {
        json!({
            "key": obj.key().unwrap_or(""),
            "size": obj.size(),
            "last_modified": obj.last_modified().map(|t| t.secs()),
        })
    }).collect();

    Ok(success_response(json!({ "files": file_list })))
}

